applications/RANSModellingApplication/custom_strategies/generic_residual_based_bossak_velocity_scalar_scheme.h

```diff
@@ -0,0 +1,127 @@
+
+    using SystemVectorType = typename BaseType::SystemVectorType;
+
+    using LocalSystemMatrixType = typename BaseType::LocalSystemMatrixType;
+
+    using LocalSystemVectorType = typename BaseType::LocalSystemVectorType;
+
+    using DofsArrayType = typename BaseType::DofsArrayType;
+
+    /// Constructor.
+
+    GenericResidualBasedBossakVelocityScalarScheme(const double AlphaBossak,
+                                                   const double RelaxationFactor,
+                                                   const Variable<double>& rScalarVariable,
+                                                   const Variable<double>& rScalarRateVariable,
+                                                   const Variable<double>& rRelaxedScalarRateVariable)
+        : ResidualBasedBossakVelocityScheme<TSparseSpace, TDenseSpace>(
+              AlphaBossak, RelaxationFactor, {}, {&rScalarVariable}, {&rScalarRateVariable}, {}, {}, {}),
+          mrScalarRateVariable(rScalarRateVariable),
+          mrRelaxedScalarRateVariable(rRelaxedScalarRateVariable)
+    {
+    }
+
+    void Update(ModelPart& rModelPart,
+                DofsArrayType& rDofSet,
+                SystemMatrixType& rA,
+                SystemVectorType& rDx,
+                SystemVectorType& rb) override
+    {
+        KRATOS_TRY;
+
+        BaseType::Update(rModelPart, rDofSet, rA, rDx, rb);
+
+        // Updating the auxiliary variables
+        const int number_of_nodes = rModelPart.NumberOfNodes();
+#pragma omp parallel for
+        for (int iNode = 0; iNode < number_of_nodes; ++iNode)
+        {
+            NodeType& r_node = *(rModelPart.NodesBegin() + iNode);
+            const double scalar_rate_dot_old =
+                r_node.FastGetSolutionStepValue(mrScalarRateVariable, 1);
+            const double scalar_rate_dot =
+                r_node.FastGetSolutionStepValue(mrScalarRateVariable, 0);
+
+            r_node.FastGetSolutionStepValue(mrRelaxedScalarRateVariable) =
+                this->mAlphaBossak * scalar_rate_dot_old +
+                (1.0 - this->mAlphaBossak) * scalar_rate_dot;
+        }
+
+        KRATOS_CATCH("");
+    }
+
+    int Check(ModelPart& rModelPart) override
+    {
+        KRATOS_TRY
+
+        int value = BaseType::Check(rModelPart);
+
+        KRATOS_CHECK_VARIABLE_KEY(mrScalarRateVariable);
+        KRATOS_CHECK_VARIABLE_KEY(mrRelaxedScalarRateVariable);
+
+        const int number_of_nodes = rModelPart.NumberOfNodes();
+#pragma omp parallel for
+        for (int iNode = 0; iNode < number_of_nodes; ++iNode)
+        {
+            NodeType& r_node = *(rModelPart.NodesBegin() + iNode);
+            KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(mrScalarRateVariable, r_node);
+            KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(mrRelaxedScalarRateVariable, r_node);
+        }
+
+        return value;
+        KRATOS_CATCH("");
+    }
+
+private:
+    Variable<double> const mrScalarRateVariable;
+    Variable<double> const mrRelaxedScalarRateVariable;
```

Comment:
There is an open issue (#3789) about problems when copying variables, could you verify that this is not a problem here? (Is this class used in the newer working examples?) If it gives problems, you could use `const*` as a workaround.

Reply:
I would also suggest to use `const *`